Fix critical proposer selection bug #4259

prestonvanloon opened the pull request:
Resolves #4259.

This is likely consensus breaking in the current testnet. 

OK to merge into 0.9.2. Testnet restart required!

terencechain commented:
Changed base to v0.9.2

I can help you resolve the conflict if needed

prestonvanloon commented:
Is this confirmed consensus breaking?

Can you sync 0.9.2 with master? Why are there conflicts?

terencechain commented:
> Is this confirmed consensus breaking?

yes, 2 nodes both have different proposer selection logic

prestonvanloon commented:
Let's wait to change base until 0.9.2 is synced. Otherwise the diff is impossible to review

prestonvanloon commented:
I did see this when trying to sync with these changes
```
[2019-12-11 15:35:40] DEBUG forkchoice: Executing state transition on block slot=115
panic: could not process blockCopy from fork choice service: could not execute state transition: could not process block: could not process block header: parent root 0xde64d3607af5b1ea9bf3cd8ff6fe0bc1baf8c27155
c61eebffa2385b747b4383 does not match the latest block header signing root in state 0xc170c1197d52b9a636c4919ddb8afd502b8685b26ef7412b189c3b17890662e2

goroutine 295 [running]:
github.com/prysmaticlabs/prysm/beacon-chain/sync/initial-sync.(*InitialSync).Start(0xc00174c540)
        beacon-chain/sync/initial-sync/service.go:132 +0xf47
created by github.com/prysmaticlabs/prysm/shared.(*ServiceRegistry).StartAll
        shared/service_registry.go:44 +0x23e
```

prestonvanloon commented:
Some other test is failing, trying to fix that.

prestonvanloon commented:
`TestServer_ListAssignments_Pagination_DefaultPageSize_FromArchive` is failing. @rauljordan can you take a look when you have a moment?

terencechain commented:
> What is this? Duplicate logic?

Duplicated logic for archival service. @rauljordan is it needed?

terencechain commented:
This is ready. Open items:
1. Found there's duplicated proposer selection logic for archival service
2. E2e testing failure

prestonvanloon commented:
This is failing e2e because 0.9.2 is failing e2e, unfortunately. So we don't know this this works at runtime.

terencechain commented:
> This is failing e2e because 0.9.2 is failing e2e, unfortunately. So we don't know this this works at runtime.

v0.9.2 doesn't work run time
`sync<->pool<->rpc` is undergoing major refactor at this moment

terencechain commented:
For this, we should add a regression test where pre computed `proposer_index` is greater than `len(active_vaidators)`,  I don't see there's one yet

prestonvanloon commented:
I will refactor ComputeProposerIndex to accept a list of validators rather than the whole state.

```go
func ComputeProposerIndex(validators []*pb.Validator, indices []uint64, seed [32]byte) (uint64, error) {
```

This would allow deduplication. I'm boarding flight now, will push the changes when the flight lands.

prestonvanloon commented:
Refactored for deduplication, added regression test in response to @terencechain feedback, added panic safety and appropriate test. 

This is ready for final review.

nisdas commented on beacon-chain/core/helpers/validators_test.go:
```diff
@@ -405,3 +405,126 @@ func TestActiveValidatorIndices(t *testing.T) {
+		want    uint64
+		wantErr bool
+	}{
+		{
```
can you also one more test case that closely mimics for what we observed in our testnet ?  example: `validator > 600` isnt selected for proposal

prestonvanloon replied:
There is one for that, I forgot to rename it, but see `second_half_active` which is the regression that only the first 5 would have been selected when the latter 5 were the only ones active.
